Lib/transform.py

```diff
@@ -0,0 +1,453 @@
+    p = np.sqrt(x*x + y*y)
+    p2 = p*p
+
+    # define a few convenient intermediate terms
+
+    q = p2*(h - e2*N + N)*(h - e2*N + N) + z*z*(h+N)*(h+N)
+    s = N*e2*sinPhi*cosPhi/(1.0 - e2*sinPhi2)
+    
+    # compute the four partials
+
+    dfdh = -1.0
+
+    dfdphi = p*sinPhi/cosPhi2 - s
+
+    dgdh = (-1.0)*e2*N*p*z/q
+
+    dgdphi = (e2*h*p*z*s/q) - 1.0
+    
+    # built J_F for output
+    J_F = [dfdh, dfdphi, dgdh, dgdphi]
+
+    return J_F
+
+#----------------------------------------------------------------------------
+
+def xyz_to_llh(x, y, z):
+    
+    """Convert cartesian coordinates to WGS84 lat., lon., and height above
+    reference ellipsoid. Cartesian coordinates must be passed in meters. 
+    Latitude and longitude are returned in degrees and height is returned in
+    meters. With:
+
+        -90 < lat. < 90     and       0 < lon. < 360
+
+    Routine starts with an initial guess assuming a spherical Earth and then
+    iterates towards the WGS84 solution using Newton's method.
+
+    Example:
+
+    >>> x, y, z = 100000, 100000, 1000000
+    >>> xyz_to_llh(x, y, z)
+    [82.27427630552833, 45.0, -5347205.038887162]
+    
+    Ref: Misra, P. & Enge, P. "Global Positioning System", 2nd ed., p. 134-135
+    """
+    # convert inputs to floats
+    x = float(x)
+    y = float(y)
+    z = float(z)
+
+    # method fails when x and y are both exactly 0; exit and return
+    # error message under this condition:
+
+    if x == 0.0 and y == 0.0:
+        exit("error:  x and y cannot both be exactly zero, point must be at least 1e-11 m from pole")
+    
+    # compute the longitude
+    preLon = np.arctan2(y,x)    # arctan2 returns angle in rads 
+                                # between [-pi, pi]
+
+    radLon = preLon%(2*np.pi)   # modulo 2*pi returns value 
+                                # in rads between [0, 2*pi)
+
+    # compute the projection in the xy-plane of the line extending 
+    # from the z-axis to the point that is  normal to the ellipsoid at
+    # the surface
+
+    p = np.sqrt(x*x + y*y)
+    
+    # Newton's method requires initial guess
+    # start with guess assuming spherical Earth
+    # with radius equal to Earth's semi-major axis
+
+    radE0 = a 
+    dist0 = np.sqrt(x*x + y*y + z*z)
+    h_prime = dist0 - radE0
+    lat_prime = np.arctan2(z,p)     # p is always positive, so even though arctan2
+                                    # returns angle in [-pi,pi] with p positive
+                                    # lat will always be in [-pi/2,pi/2]
+
+    # compute misfit between actual and initial guess
+    
+    f_prime, g_prime = compute_F(h_prime, lat_prime, x, y, z)
+
+    # while the error in h and phi is greater than 1e-9 meters
+    # keep iterating toward a better solution (note: latitude error 
+    # (i.e. g_prime) is still in radians at this point. And 1e-12 
+    # radians of latitude is about 1e-9 meters)
+    
+    while np.absolute(f_prime) > 1e-9 or np.absolute(g_prime) > 1e-12:
```

Comment:
Is there a possibility this could loop forever? Should you add a counter that exits with an error if its looped more than like a billion times?

Reply:
Very very unlikely but you are probably right, should add a max iteration condition